### ISSUE TITLE
Speed improvements

### ISSUE THREAD

**Issue by Wessie** [worldveil/dejavu#24]
Are there any possible tweaks that could be done to the algorithm to improve performance. 

dejavu currently seems to net about x4 real-time fingerprinting speeds.

I've been testing with lower `sample rate`, different `window size`, `overlap ratio` and I think there are certainly speedups to be found if we can find the sweet spot between speed and accuracy (None of my tests have acceptable accuracy as of yet, and varying speedups).

Since I'm not too familiar with the exact algorithm I thought it might be a better idea to involve everyone in it.

**Comment by Hoohaha** [comment]
I am also researching at the audio recognition. But it is big different with the part of fingerprints-extract. I think that the speed is the most point. With the numbers of songs increase in database, the time cost will be terrible, and the users would not tolerate this.  Actually, the speed bottleneck is the progress of searching in the database.  If we can find a new music index or a new way to compute the distance between two songs, the speed will be greatly improved.

**Comment by worldveil** [comment]
@Hoohaha fingerprinting or locally sensitive hashing in general does not produce distances between audio files. It merely creates derivative signals that are deterministic and noise-resistant.

**Comment by zzzhe1990** [comment]
sorry for digging this old post out. I am now working on accelerating the training part, actually the fingerprinting. I got a big sample of about 300 songs. I noticed that mysql takes an extremely large of time that made me crazy. By typing "top", I can see that my multi-processing "python" task have done, but "mysql" task keeps running for hours (at around 40-50% cpu usage) as well as the master process of "python" at cpu usage of 4-7%. So, this is telling me that mysql is the bottleneck on fingerprinting for a large size of samples. Is that the same on your side? Do you have any experience of running test on large samples? 

Thanks for any information.

**Comment by zzzhe1990** [comment]
Follow up. The problem I mentioned above might be caused by MySQL, but there is another possibility that something wrong with the multi-processes termination. In __init__.py, there is a code segment like:

`while True:
            try:
                song_name, hashes, file_hash = iterator.next()
            except multiprocessing.TimeoutError:
                continue
            except StopIteration:
                break
            except:
                print("Failed fingerprinting")
                # Print traceback because we can't reraise it here                                                                          
                traceback.print_exc(file=sys.stdout)
            else:
                sid = self.db.insert_song(song_name, file_hash)

                self.db.insert_hashes(sid, hashes)
                self.db.set_song_fingerprinted(sid)
                self.get_fingerprinted_songs()
        pool.close()
        pool.join()
`
Since I made changes to scipy ni_filter.c to enable multi-threading, which gives me a fly on fingerprinting, the situation now is that python runs multi-processing, and each process has multi-threads supported. I do not know if the support of multi-threading cause any troubles in the code above? Like the termination of some thread is not captured, so that the termination loop will never stop?

**Comment by AforAnonymous** [comment]
How about using the Constant-Q transform to speed this up?

See https://en.wikipedia.org/wiki/Constant-Q_transform#Fast_calculation_using_FFT


(for more related & interesting reading, see here, although I wouldn't expect speed improvements via that, it's certainly interesting:

- https://en.wikipedia.org/wiki/Morlet_wavelet
- https://en.wikipedia.org/wiki/List_of_Fourier-related_transforms
- https://en.wikipedia.org/wiki/Nyquist%E2%80%93Shannon_sampling_theorem#Nonuniform_sampling
- https://www.wired.com/2010/02/ff_algorithm/all/1
- http://nuit-blanche.blogspot.com/2010/03/why-compressed-sensing-is-not-csi.html
)

**Comment by gudlucsam** [comment]
Hi @worldveil I tried implementing the fingerprinting in Cython, but didn't get most improvement.... after implementing i profiled it and realized that execution time only reduced by about 1minute.
Also, since Scipy is already in C, there wasn't much improvement i could make on it, but only to statically type variables. Are there ideas of any algorithm for peak finding that you could suggest? 
thank you.

**Comment by milahu** [comment]
timestretch filter is a "simple but effective" way, to speed up both hashing and matching.

for example ....

    ffmpeg -i input.mp4 -map 0:a:0 -ac 1 -filter:a atempo=2.0 -c:a pcm_s16le output.wav

.... to make audio shorter by factor 2

for better accuracy, you can do "two-stage testing", 
with a "quick but dirty" first stage, and a "slow but strict" second stage.
[ the game of Greed and Love .... ]
this will make hashing slower, but matching faster.

another optimization path:
stop matching at the first "good match", and avoid scanning the rest of the haystack.

**Comment by NathanielCustom** [comment]
The first parts go over my head without deeper explanation, so I can't comment.
The last part on "stop matching at the first "good match"" I want to explore...

What defines a "good match"? From my understanding Dejavu takes the sample and tries to find all the matches within each neighborhood slice for each song in the database. The more matches within the neighborhood the higher the returned "confidence" value.
I currently do this in a project where I will only accept results if the "confidence" is higher than a predetermined number, but AFTER the database has been searched. To apply this logic while the database is being searched is something I will have to look into, but I imagine that the value used to define a "good match" will depend on the project or type of audio fingerprinted.

**Comment by milahu** [comment]
>> stop matching at the first "good match"

> What defines a "good match"?

let me rephrase to:
add support for "inter process communication" and "parallel processing",
to allow better control / flexibility / speed

the matching process should "yield" matches as soon as possible,
and continue to "yield" confidence values for the current match.

the controller process does observe the match/confidence data,
and when "satisfied" he can stop the matching process.
this is useful to speed up "single file matching".

.... or trigger a callback function, 
like "print match to logfile" or "cut stream from last match to current match" or "mute commercial break",
and continue to observe the matching process.
this is useful to analyze "live streams"
or to observe "long streams" for different audio patterns
----> simple "machine hearing"

**Comment by artgancz** [comment]
There are two pretty easy tricks to speed up this project a little bit, first one: replace matplotlib.specgram inside fingerprint.py with scipy.signal.spectrogram (with detrend=False for compatibility). Second by tuning MySQL: https://www.percona.com/blog/2006/09/29/what-to-tune-in-mysql-server-after-installation/

**Comment by omertoptas** [comment]
> Unfortunately I focused mostly on the speed of fingerprinting and searching in database, so if you are interested in this part of the project then I can give some more tips

Hello again , I need your tips now , I am struggling while searching in database, please give me some advice, thank you.

**Comment by mauriciorepetto** [comment]
Hello everyone, I've added a pr https://github.com/worldveil/dejavu/pull/205 with some modifications to the code, I could get down recognizing times from 20 seconds to ~3 seconds on songs of 3:30 minutes in average without hurting the accuracy. I've the same doubts of you regarding the erosion portion but I didn't want to discard it since does not take much time to process. 
The pr is actually to migrate dejavu to python 3.6.6, solving a couple of bugs that I identified, adding support for mysql new connector and postgresql. There a couple of features more as well.. but please feel free to try it and let me know what do you think, thanks.
